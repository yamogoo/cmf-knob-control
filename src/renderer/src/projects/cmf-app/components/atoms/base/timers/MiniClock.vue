<template lang="pug">
div(:class="[`${APP_PREFIX}-mini-clock`]")
  span(
    data-test-id="mini-clock__time"
  ) {{ time }}
</template>

<script setup lang="ts">
import { APP_PREFIX } from '@app/config'

import { useCurrentTime } from '@app/composables/useTime'

const time = useCurrentTime()
</script>

<style lang="scss">
.#{$APP_PREFIX}-mini-clock {
  font-size: 21px;
  color: $WHITE;

  span {
    color: inherit;
    font-size: inherit;
  }
}
</style>
