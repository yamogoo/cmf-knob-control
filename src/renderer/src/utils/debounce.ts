export function debounce<T extends (...args: any[]) => void>(
  func: T,
  ms: number
) {
  let timerId: ReturnType<typeof setTimeout> | null = null

  return function <U>(this: U, ...args: Parameters<typeof func>) {
    const ctx = this
    if (timerId) clearTimeout(timerId)
    timerId = setTimeout(() => {
      func.apply(ctx, args)
    }, ms)
  }
}
