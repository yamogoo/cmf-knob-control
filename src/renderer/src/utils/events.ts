export const createCustomEvent = <T extends keyof GlobalEventHandlersEventMap>(
  type: T,
  eventInitDict: CustomEventInit<
    GlobalEventHandlersEventMap[T] extends CustomEvent<infer T> ? T : never
  >
) => new CustomEvent(type, eventInitDict)
