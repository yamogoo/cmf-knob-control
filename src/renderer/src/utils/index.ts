export * from './debounce'
export * from './timer'
export * from './events'
export * from './time'
