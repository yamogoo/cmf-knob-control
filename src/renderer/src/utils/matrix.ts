import type { Matrix3D } from '@typings/index'

export const flattenMatrix = <T>(_matrix: Readonly<Matrix3D<T>>): Array<T> => {
  const matrix: Matrix3D<T> = [..._matrix]
  const flatten: Array<T> = []

  for (let y = 0; y < matrix.length; y++) {
    for (let x = 0; x < matrix[y].length; x++) {
      flatten.push(matrix[y][x])
    }
  }

  return flatten
}
