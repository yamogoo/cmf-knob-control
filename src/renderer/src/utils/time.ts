export const getCurrentTime = (): string => {
  const d = new Date()
  return `${('0' + d.getHours()).slice(-2)}:${('0' + d.getMinutes()).slice(-2)}`
}
