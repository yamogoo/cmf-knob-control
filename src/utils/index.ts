export * from './debounce'
export * from './timer'
export * from './events'
