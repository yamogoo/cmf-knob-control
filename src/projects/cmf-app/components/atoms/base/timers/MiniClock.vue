<template lang="pug">
div(:class="[`${APP_PREFIX}-mini-clock`]")
    span {{ time }}
</template>

<script setup lang="ts">
import { cancelableInterval } from '@/utils';
import { APP_PREFIX } from '@app/config';
import { onMounted, onUnmounted, ref } from 'vue';

const time = ref('--:--');

let timer: ReturnType<typeof cancelableInterval> | undefined = undefined;

const getTime = (): void => {
    const d = new Date();
    time.value = `${('0' + d.getHours()).slice(-2)}:${('0' + d.getMinutes()).slice(-2)}`;
};

onMounted(() => {
    getTime();
    timer = cancelableInterval(() => {
        getTime();
    }, 1000);
});

onUnmounted(() => {
    if (timer) timer();
});
</script>

<style lang="scss">
.#{$APP_PREFIX}-mini-clock {
    font-size: 21px;
    color: $WHITE;

    span {
        color: inherit;
        font-size: inherit;
    }
}
</style>