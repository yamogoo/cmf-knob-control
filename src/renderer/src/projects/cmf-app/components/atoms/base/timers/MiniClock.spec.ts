import { describe, expect, test } from 'vitest'
import { mount } from '@vue/test-utils'

import { getCurrentTime } from '~@/utils'
import MiniClock from './MiniClock.vue'

describe('MiniClock', () => {
  const currentTime = getCurrentTime()

  test(`should render current time (${currentTime})`, () => {
    const component = mount(MiniClock)

    const timeEl = component.find('[data-test-id="mini-clock__time"]').text()
    expect(timeEl).toContain(String(currentTime))
    expect(timeEl).toMatchSnapshot()
  })
})
