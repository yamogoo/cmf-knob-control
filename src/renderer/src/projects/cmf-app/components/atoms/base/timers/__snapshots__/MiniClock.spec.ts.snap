// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`MiniClock > should render current time (10:13) 1`] = `"10:13"`;

exports[`MiniClock > should render current time (10:14) 1`] = `"10:14"`;

exports[`MiniClock > should render current time (10:15) 1`] = `"10:15"`;

exports[`MiniClock > should render current time (10:36) 1`] = `"10:36"`;

exports[`MiniClock > should render current time (10:38) 1`] = `"10:38"`;

exports[`MiniClock > should render current time (11:17) 1`] = `"11:17"`;

exports[`MiniClock > should render current time (11:18) 1`] = `"11:18"`;

exports[`MiniClock > should render current time (11:19) 1`] = `"11:19"`;

exports[`MiniClock > should render current time (11:20) 1`] = `"11:20"`;

exports[`MiniClock > should render current time (11:21) 1`] = `"11:21"`;

exports[`MiniClock > should render current time (11:24) 1`] = `"11:24"`;

exports[`MiniClock > should render current time (11:25) 1`] = `"11:25"`;

exports[`MiniClock > should render current time (11:26) 1`] = `"11:26"`;

exports[`MiniClock > should render current time (12:36) 1`] = `"12:36"`;

exports[`MiniClock > should render current time (12:38) 1`] = `"12:38"`;

exports[`MiniClock > should render current time (12:40) 1`] = `"12:40"`;

exports[`MiniClock > should render current time (20:15) 1`] = `"20:15"`;

exports[`MiniClock > should render current time (20:28) 1`] = `"20:28"`;

exports[`MiniClock > should render current time (20:29) 1`] = `"20:29"`;

exports[`MiniClock > should render current time (20:30) 1`] = `"20:30"`;

exports[`MiniClock > should render current time (20:51) 1`] = `"20:51"`;

exports[`MiniClock > should render current time (20:53) 1`] = `"20:53"`;
